motor/aiohttp/__init__.py

```diff
@@ -190,7 +191,19 @@ async def __call__(self, request):
             raise aiohttp.web.HTTPNotFound(text=request.path)
 
         resp = aiohttp.web.StreamResponse()
-        self._set_standard_headers(request.path, resp, gridout)
+
+        # Calculate the sha256 checksum for the GridFS file
+        # for a FIPS-compliant Etag HTTP header.
+        sha = hashlib.sha256()
+        while True:
+            chunk = await gridout.readchunk()
+            if not chunk:
+                break
+            sha.update(chunk)
+        gridout.seek(0)
+        sha = sha.hexdigest()
```

Comment:
We can't merge this change since it makes this method read the entire file twice. It also still reads the entire file even just to return a HTTP 304 Not Modified code.  https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag says:

> The ETag (or entity tag) HTTP response header is an identifier for a specific version of a resource. It lets caches be more efficient and save bandwidth, as a web server does not need to resend a full response if the content was not changed. Additionally, etags help to prevent simultaneous updates of a resource from overwriting each other ("mid-air collisions").
> If the resource at a given URL changes, a new Etag value must be generated. A comparison of them can determine whether two representations of a resource are the same. Etags are therefore similar to fingerprints, and might also be used for tracking purposes by some servers. They might also be set to persist indefinitely by a tracking server.

I propose to hash the "Files collection document" or even a subset of the fields instead. GridFS files are supposed to be immutable so we don't have to worry about the contents changing. The only issue is if a user deletes and re-uploads a gridfs file with the same _id. In that case other fields in the files document (like uploadDate and length) would most likely change as well:
```
{
  "_id" : <TFileId>,
  "length" : <Int64>,
  "chunkSize" : <Int32>,
  "uploadDate" : <BSON datetime, ms since Unix epoch in UTC>,
  "md5" : <hex string>,
  "filename" : <string>,
  "contentType" : <string>,
  "aliases" : <string array>,
  "metadata" : <Document>
}
```
https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst

You can hash these fields on the GridOut:
```
    _id = _grid_out_property("_id", "The ``'_id'`` value for this file.")
    filename = _grid_out_property("filename", "Name of this file.")
    name = _grid_out_property("filename", "Alias for `filename`.")
    content_type = _grid_out_property("contentType", "Mime-type for this file.")
    length = _grid_out_property("length", "Length (in bytes) of this file.")
    chunk_size = _grid_out_property("chunkSize", "Chunk size for this file.")
    upload_date = _grid_out_property("uploadDate",
                                     "Date that this file was first uploaded.")
    aliases = _grid_out_property("aliases", "List of aliases for this file.")
    metadata = _grid_out_property("metadata", "Metadata attached to this file.")
    md5 = _grid_out_property("md5", "MD5 of the contents of this file "
                             "if an md5 sum was created.")
```

It might even be enough to hash only _id+length+upload_date. What do you think?

Reply:
Sounds reasonable to me